src/Castle.Core/DynamicProxy/ModuleScope.cs

```diff
@@ -174,11 +189,94 @@ public Type GetFromCache(CacheKey key)
 		/// </summary>
 		/// <param name = "key">The key to be associated with the type.</param>
 		/// <param name = "type">The type to be stored in the cache.</param>
+		[Obsolete]
+		[EditorBrowsable(EditorBrowsableState.Never)]
 		public void RegisterInCache(CacheKey key, Type type)
 		{
 			typeCache[key] = type;
 		}
 
+		/// <summary>
+		///   Returns a type from this scope's type cache, or registers it in the cache if the key cannot be found.
+		/// </summary>
+		/// <param name="key">The key to be looked up in the cache.</param>
+		/// <param name="valueFactory">A function producing the type to be registered in the cache.</param>
+		/// <returns>The type from this scope's type cache matching the key, or null if the key cannot be found</returns>
+		/// <remarks>
+		///   This method does not synchronizes accesses to this scope's type cache using read/write locks.
+		///   Only use this method when you know for sure that a method further up in the call stack already holds
+		///   a write lock. The function must not access this scope's type cache.
+		/// </remarks>
+		public Type GetFromOrRegisterInCache(CacheKey key, Func<CacheKey, Type> valueFactory)
+		{
+			Type value;
+
+			Debug.Assert(key != null);
+			Debug.Assert(valueFactory != null);
+
+			if (this.typeCache.TryGetValue(key, out value))
+			{
+				return value;
+			}
+			else
+			{
+				value = valueFactory.Invoke(key);
+				this.typeCache.Add(key, value);
+				return value;
+			}
+		}
+
+		/// <summary>
+		///   Returns a type from this scope's type cache, or registers it in the cache if the key cannot be found.
+		/// </summary>
+		/// <param name="key">The key to be looked up in the cache.</param>
+		/// <param name="valueFactory">A function producing the type to be registered in the cache.</param>
+		/// <returns>The type from this scope's type cache matching the key, or null if the key cannot be found</returns>
+		/// <remarks>
+		///   This method synchronizes accesses to this scope's type cache using read/write locks.
+		///   If the specified key cannot be found, the factory function is invoked while a write lock is held.
+		///   The function must not access this scope's type cache.
+		/// </remarks>
+		public Type GetFromOrRegisterInCacheSynchronized(CacheKey key, Func<CacheKey, Type> valueFactory)
```

Comment:
I would have thought this method would be named `GetFromOrRegisterInCache` as the go to one (like how `AddCacheMappings` has built-in synchronisation), and the previous one would be suffixed to indicate it is the special one, e.g. `GetFromOrRegisterInCacheHoldingLock` or something.

On a separate note how do users use the first one without synchronisation without using the deprecated `Lock` property?

Maybe `GetOrAddToCache` would be a simpler name.

Reply:
@jonorossi:

> On a separate note how do users use the first one without synchronisation without using the deprecated `Lock` property?

If they need locking, they use the other, synchronized method. If they don't need locking, then they don't need `Lock` in the first place.

The question here is probably, who are the users you're thinking of? Why should anyone even try to update the proxy type cache? Why is the proxy type cache, along with `CacheKey`, even exposed publicly?

> Maybe `GetOrAddToCache` would be a simpler name.

I tried to stay close to what was there before the changes, but I agree.